source/base/parameter_handler.cc

```diff
@@ -1760,155 +1772,110 @@ namespace
+            // match the ones in the 'destination' tree
           }
         else if (p->second.get_optional<std::string>("alias"))
           {
-            // it is an alias node. alias nodes are static and
-            // there is nothing to do here (but the same applies as
-            // mentioned in the comment above about the static
-            // nodes inside parameter nodes
+            // it is an alias node. alias nodes are static and there is
+            // nothing to do here (but the same applies as mentioned in the
+            // comment above about the static nodes inside parameter nodes
           }
         else
           {
             // it must be a subsection
-            const bool result
-              = read_xml_recursively (p->second,
-                                      (current_path == "" ?
-                                       p->first :
-                                       current_path + path_separator + p->first),
-                                      path_separator,
-                                      patterns,
-                                      destination);
-
-            // see if the recursive read
-            // succeeded. if yes, continue,
-            // otherwise exit now
-            if (result == false)
-              return false;
+            read_xml_recursively (p->second,
+                                  (current_path == "" ?
+                                   p->first :
+                                   current_path + path_separator + p->first),
+                                  path_separator,
+                                  patterns,
+                                  destination);
           }
       }
-
-    return true;
   }
 }
 
 
 
-bool ParameterHandler::read_input_from_xml (std::istream &in)
+bool ParameterHandler::read_input_from_xml(std::istream &in)
+{
+  parse_input_from_xml(in);
+  return true;
+}
+
+
+
+void ParameterHandler::parse_input_from_xml (std::istream &in)
 {
   AssertThrow(in, ExcIO());
   // read the XML tree assuming that (as we
   // do in print_parameters(XML) it has only
   // a single top-level node called
   // "ParameterHandler"
   boost::property_tree::ptree single_node_tree;
-  try
-    {
-      read_xml (in, single_node_tree);
-    }
-  catch (...)
-    {
-      std::cerr << "This input stream appears not to be valid XML"
-                << std::endl;
-      return false;
-    }
+  // This boost function will raise an exception if this is not a valid XML
+  // file.
+  read_xml (in, single_node_tree);
 
-  // make sure there is a top-level element
+  // make sure there is a single top-level element
   // called "ParameterHandler"
-  if (!single_node_tree.get_optional<std::string>("ParameterHandler"))
-    {
-      std::cerr << "There is no top-level XML element called \"ParameterHandler\"."
-                << std::endl;
-      return false;
-    }
+  AssertThrow (single_node_tree.get_optional<std::string>("ParameterHandler"),
```

Comment:
To be honest I just got lazy by the end and stopped creating new exception classes. I'll fix this.